### ISSUE TITLE
helpers.isReadStream identifies only fs-created streams

### ISSUE THREAD

**Issue by jackwanders** [request/request#1949]
https://github.com/request/request/blob/master/lib/helpers.js#L41

`.isReadStream` checks to see if the provided object has three truthy properties: `.readable`, `.path`, and `.mode`. While these three properties all exist on streams created by [`fs.createReadStream`](https://nodejs.org/api/fs.html#fs_fs_createreadstream_path_options), this is not true for all readable streams that may contain file data.

In particular, I am working on a project using `sails@0.11.3` and `skipper@0.5.8`. The "upstream" paradigm that `skipper` uses creates a transform stream containing file data. This allows `sails` users to write file management middleware that sends file data to any desired storage location without first writing any data to disk. Because the file data is coming directly from the request and into the transform stream, no file system read is happening, and therefore no `.path` and `.mode` properties exist for these streams which contain file data.

I'm opening this issue to ask whether there is a more flexible way for `request` to identify readable streams. Node exposes four different stream classes that implement the readable interface, `Readable`, `Duplex`, `Transform`, and `PassThrough`. All of these stream types appear to share two properties:

```javascript
['Readable', 'Transform', 'Duplex', 'PassThrough'].forEach(function(type) {
  var stream = new require('stream')[type]();
  console.log(type, !!stream._readableState, !!stream.readable);
});

/* Output
Readable true true
Transform true true
Duplex true true
PassThrough true true
*/
```

One possible refactor that appears to allow `request` to accept non fs file streams while avoiding a regression is:

```javascript
function isReadStream (rs) {
  return rs.readable && ((rs.path && rs.mode) || rs._readableState);
}
```

Are there any foreseeable downsides to opening `request` up to accepting any readable stream, as opposed to only those created by `fs`? Is there a catastrophic failure point somewhere where `request` is assuming the stream to be an `fs` stream?

**Comment by jackwanders** [comment]
Another option would be to check to see if the provided `rs` object is an instance of `stream.Stream`:

```javascript
var stream = require('stream');
['Readable', 'Duplex', 'Transform', 'PassThrough', 'Writable'].forEach(function(type) {
  var rs = stream[type]();
  console.log(type, rs.readable, rs instanceof stream.Stream);
});

/* Output
Readable true true
Duplex true true
Transform true true
PassThrough true true
Writable undefined true
*/
```

**Comment by simov** [comment]
Hi @jackwanders thanks for the thorough bug report!

Request already depends on [isstream](https://www.npmjs.com/package/isstream). Do you think that would work for all of the cases where the helper function is being used?

If that's the case you can submit a patch about it.

**Comment by jackwanders** [comment]
Good point, @simov. In that case, it seems relatively straightforward to just swap the `isReadStream` method used within `request.js`.

I've submitted a pull request to that effect: https://github.com/request/request/pull/1964

**Comment by jackwanders** [comment]
I've closed #1964 and likewise, I'm going to close this issue, as my particular problem was unrelated to how `request` detects streams, but rather that I needed to specify the size of the file stream I provided to the `formData` has within my call to `request.post`.
